x-pack/plugins/fleet/server/services/epm/registry/index.ts

```diff
@@ -256,32 +282,67 @@ function ensureContentType(archivePath: string) {
   return contentType;
 }
 
-export async function ensureCachedArchiveInfo(
-  name: string,
-  version: string,
-  installSource: InstallSource = 'registry'
-) {
-  const paths = getArchiveFilelist({ name, version });
-  if (!paths || paths.length === 0) {
-    if (installSource === 'registry') {
-      await getRegistryPackage(name, version);
-    } else {
-      throw new PackageCacheError(
-        `Package ${name}-${version} not cached. If it was uploaded, try uninstalling and reinstalling manually.`
-      );
+export async function fetchArchiveBuffer({
+  pkgName,
+  pkgVersion,
+  verify,
+  ignoreUnverified = false,
+}: {
+  pkgName: string;
+  pkgVersion: string;
+  verify: boolean;
+  ignoreUnverified?: boolean;
+}): Promise<{
+  archiveBuffer: Buffer;
+  archivePath: string;
+  verificationResult?: PackageVerificationResult;
+}> {
+  const logger = appContextService.getLogger();
+  const { download: archivePath } = await getInfo(pkgName, pkgVersion);
+  const archiveUrl = `${getRegistryUrl()}${archivePath}`;
+  const archiveBuffer = await getResponseStream(archiveUrl).then(streamToBuffer);
+  if (verify) {
+    const verificationResult = await verifyPackageArchiveSignature({
+      pkgName,
+      pkgVersion,
+      pkgArchiveBuffer: archiveBuffer,
+      logger,
+    });
+
+    if (verificationResult.verificationStatus === 'unverified' && !ignoreUnverified) {
+      throw new PackageFailedVerificationError(`${pkgName}-${pkgVersion}`);
     }
+    return { archiveBuffer, archivePath, verificationResult };
   }
+  return { archiveBuffer, archivePath };
 }
 
-export async function fetchArchiveBuffer(
-  pkgName: string,
-  pkgVersion: string
-): Promise<{ archiveBuffer: Buffer; archivePath: string }> {
-  const { download: archivePath } = await getInfo(pkgName, pkgVersion);
-  const archiveUrl = `${getRegistryUrl()}${archivePath}`;
-  const archiveBuffer = await getResponseStream(archiveUrl).then(streamToBuffer);
+export async function getPackageArchiveSignatureOrUndefined({
+  pkgName,
+  pkgVersion,
+  logger,
+}: {
+  pkgName: string;
+  pkgVersion: string;
+  logger: Logger;
+}): Promise<string | undefined> {
+  const { signature_path: signaturePath } = await getInfo(pkgName, pkgVersion);
+
+  if (!signaturePath) {
+    logger.debug(
+      `Package ${pkgName}-${pkgVersion} does not have a signature_path, verification will not be possible.`
```

Comment:
This debug log should make the warning above. Also should they both be warnings?